store/store_test.go

```diff
@@ -2582,6 +2582,54 @@ func (t *remoteRepoTestSuite) TestUbuntuStoreRepositoryListRefreshRetryOnEOF(c *
 	c.Assert(results[0].Name(), Equals, "hello-world")
 }
 
+func (t *remoteRepoTestSuite) TestUbuntuStoreUnexpectedEOFhandling(c *C) {
+	mockServerHelper := func(failures int) error {
+		var mockServer *httptest.Server
+		n := 0
+		mockServer = httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+			n++
+			if n < failures {
+				w.Header().Add("Content-Length", "1000")
+				return
+			}
+			io.WriteString(w, MockUpdatesJSON)
+		}))
+
+		c.Assert(mockServer, NotNil)
+		defer mockServer.Close()
+
+		var err error
+		bulkURI, err := url.Parse(mockServer.URL + "/updates/")
+		c.Assert(err, IsNil)
+		cfg := Config{
+			BulkURI: bulkURI,
+		}
+		authContext := &testAuthContext{c: c, device: t.device}
+		repo := New(&cfg, authContext)
+		c.Assert(repo, NotNil)
+
+		_, err = repo.ListRefresh([]*RefreshCandidate{
+			{
+				SnapID:   helloWorldSnapID,
+				Channel:  "stable",
+				Revision: snap.R(1),
+				Epoch:    "0",
+			},
+		}, nil)
+		return err
+	}
+
+	// Check that we really recognize unexpected EOF error by failing on all retries
+	err := mockServerHelper(1000000)
```

Comment:
I wanted to make it infinity, so just picked a number that's closer to inifinity than - say - 10 ;). But seriously, it just needs to be grater than the number of retries we have in the store code and we will actually fail after 5 retries.